src/OpenTelemetry.Exporter.Geneva/TLDExporter/TLDTraceExporter.cs

```diff
@@ -0,0 +1,309 @@
+// <copyright file="TLDTraceExporter.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Exporter.Geneva.TLDExporter;
+
+internal sealed class TLDTraceExporter : TLDExporter, IDisposable
+{
+    private readonly string partAName = "Span";
+    private readonly byte partAFieldsCount = 3; // At least three fields: time, ext_dt_traceId, ext_dt_spanId
+    private readonly IReadOnlyDictionary<string, object> m_customFields;
+    private readonly Tuple<byte[], byte[]> repeatedPartAFields;
+    private static readonly string INVALID_SPAN_ID = default(ActivitySpanId).ToHexString();
+
+    private readonly EventProvider eventProvider;
+    private static readonly ThreadLocal<EventBuilder> eventBuilder = new(() => new(UncheckedASCIIEncoding.SharedInstance));
+    private static readonly ThreadLocal<List<KeyValuePair<string, object>>> keyValuePairs = new(() => new());
```

Comment:
This is used for serializing ActivityLinks and `env_properties`. I chose a `List` here instead of an array as we don't know how many such fields there could be.

Reply:
What's the initial capacity?

Reply:
> What's the initial capacity?

I let it use the default behavior.

> (and what's the reallocation behavior? e.g. would this give OOM?)

I understand that this might lead to unbounded memory usage and possibly even lead to OOM in extreme cases but I didn't know if we could enforce some sort of limitation on the number of links or env_properties allowed. It's straightforward for PartC fields as the TLD spec enforces a limit on the number of fields allowed. In this case though, links or env_properties would only count towards one field. If we could enforce a similar limitation here, we could very well use an array and just tell the user that they have too many links or env_properties.

Could we enforce such a limitation? If not, I guess we would have to go back to the two iterations approach.

Reply:
Added a TODO for this.